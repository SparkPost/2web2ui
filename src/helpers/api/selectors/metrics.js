export const selectRecipientDomains = ({ domains = [] } = {}) =>
  domains.map(value => ({
    type: 'domains',
    value,
  }));

export const selectSubaccounts = (subaccounts = []) =>
  subaccounts.map(({ name, id }) => ({
    type: 'subaccounts',
    value: `${name} (ID ${id})`,
    id,
  }));

export const selectSendingDomains = (sendingDomains = []) =>
  sendingDomains.map(({ domain }) => ({
    type: 'sending_domains',
    value: domain,
  }));

export const selectCampaigns = ({ campaigns = [] } = {}) =>
  campaigns.map(value => ({
    type: 'campaigns',
    value,
  }));

export const selectSubjectCampaigns = ({ 'subject-campaigns': subjectCampaigns = [] } = {}) =>
  subjectCampaigns.map(value => ({
    type: 'subject_campaigns',
    value,
  }));

export const selectSendingIps = (data = {}) => {
  const sendingIps = data['sending-ips'] || [];
  return sendingIps.map(value => ({
    type: 'sending_ips',
    value,
  }));
};

export const selectIpPools = (data = {}) => {
  const ipPools = data['ip-pools'] || [];
  return ipPools.map(value => ({
    type: 'ip_pools',
    value,
  }));
};

export const selectTemplates = ({ templates = [] } = {}) =>
  templates.map(value => ({
    type: 'templates',
    value,
  }));
