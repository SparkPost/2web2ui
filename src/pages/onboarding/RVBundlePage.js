import React from 'react';
import { CheckCircle, Send } from '@sparkpost/matchbox-icons';
import {
  Card,
  CenteredLogo,
  RecipientValidationPriceTable,
  CardTitle,
  CardContent,
} from 'src/components';
import { PageLink } from 'src/components/links';
import { Button, Modal, Panel, Stack, UnstyledLink } from 'src/components/matchbox';
import { Heading } from 'src/components/text';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import OGStyles from './RVBundlePage.module.scss';
import HibanaStyles from './RVBundlePageHibana.module.scss';

const RVPriceModal = ({ open, onClose }) => (
  <Modal.LEGACY
    open={open}
    onClose={onClose}
    showCloseButton
    title="Onboarding Recipient Validation Pricing Modal"
  >
    <Panel.LEGACY sectioned title="Pay-As-You-Go Pricing">
      <p>
        We have a monthly pay-as-you-go plan using tiered pricing. The more you validate, the less
        you pay per message.
      </p>
      <RecipientValidationPriceTable />
    </Panel.LEGACY>
  </Modal.LEGACY>
);

const RVBundlePage = () => {
  const styles = useHibanaOverride(OGStyles, HibanaStyles);
  const [modalOpen, setModalOpen] = React.useState(false);

  return (
    <>
      <div className={styles.Header}>
        <Stack>
          <CenteredLogo />
          <Heading as="h1" looksLike="h2">
            You're Almost There...
          </Heading>
          <p>
            Here at SparkPost we do it all! As part of our Recipient Validation service we also
            provide a SparkPost sending account. Make every email count and start verifying.
          </p>
        </Stack>
      </div>
      <Panel.LEGACY>
        <Panel.LEGACY.Section>
          <Stack>
            <p>
              Take a look at what's included in our Recipient Validation service and then start
              validating.
            </p>
            <div className={styles.Grid}>
              <Card>
                <CardTitle level="2">
                  <CheckCircle className={styles.ValidationIcon} size={22} />
                  Recipient Validation
                </CardTitle>
                <CardContent>
                  <Stack>
                    <p>
                      With our monthly pay-as-you-go plan using tiered pricing, the more you
                      validate, the less you pay per message.
                    </p>
                    <p>
                      <UnstyledLink
                        title="Opens a pricing dialog"
                        role="button"
                        href="#"
                        onClick={event => {
                          event.preventDefault();
                          setModalOpen(true);
                        }}
                      >
                        See Pricing
                      </UnstyledLink>
                    </p>
                  </Stack>
                </CardContent>
              </Card>
              <Card>
                <CardTitle level="2">
                  <Send className={styles.SendIcon} size={22} />
                  Email Messaging
                </CardTitle>
                <CardContent>
                  <Stack>
                    <p>
                      A full featured FREE test account is standard with all SparkPost accounts:
                    </p>
                    <ul>
                      <li>500 emails/month included for testing.</li>
                      <li>Access to all our powerful API features.</li>
                      <li>Free technical support to get you up and running.</li>
                    </ul>
                  </Stack>
                </CardContent>
              </Card>
            </div>
          </Stack>
        </Panel.LEGACY.Section>
        <Panel.LEGACY.Section>
          <PageLink as={Button} variant="primary" to="/recipient-validation">
            Start Validating
          </PageLink>
        </Panel.LEGACY.Section>
      </Panel.LEGACY>
      <RVPriceModal open={modalOpen} onClose={() => setModalOpen(false)} />
    </>
  );
};

export default RVBundlePage;
