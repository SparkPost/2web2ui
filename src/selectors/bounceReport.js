import { createSelector, createStructuredSelector } from 'reselect';
import { getBandTypes, reshapeCategories, formatAggregates } from 'src/helpers/bounce';
import { selectReportSearchOptions } from './reportSearchOptions';
import _ from 'lodash';

const selectReportOptions = state => state.reportOptions;
const selectChartLoading = ({ bounceReport }) =>
  bounceReport.aggregatesLoading || bounceReport.categoriesLoading;
const selectReasonsLoading = ({ bounceReport }) => bounceReport.reasonsLoading;
const selectTableLoading = createSelector(
  [selectChartLoading, selectReasonsLoading],
  (chartLoading, reasonsLoading) => chartLoading || reasonsLoading,
);
const selectAggregates = ({ bounceReport }) => bounceReport.aggregates;
const selectClassifications = state => _.get(state, 'bounceReport.classifications', []);
export const selectReasons = ({ bounceReport }) =>
  bounceReport.reasons.filter(reason => reason.count_bounce > 0);
const selectAdminReasons = ({ bounceReport }) =>
  bounceReport.adminReasons.filter(reason => reason.count_admin_bounce > 0);

export const selectFormattedAggregates = createSelector([selectAggregates], aggregates => {
  if (!aggregates.count_bounce && !aggregates.count_admin_bounce) {
    return [];
  }
  return formatAggregates(aggregates);
});

const selectReshapedCategories = createSelector([selectClassifications], classifications =>
  reshapeCategories(classifications),
);

const selectCategories = createSelector([selectReshapedCategories], categories =>
  _.filter(categories, ({ name }) => name !== 'Admin'),
);

const selectAdminCategories = createSelector([selectReshapedCategories], categories => {
  const adminBounces = _.find(categories, ({ name }) => name === 'Admin');
  if (adminBounces) {
    return adminBounces.children;
  }
  return [];
});

const selectBandTypes = createSelector([selectFormattedAggregates], formattedAggregates =>
  getBandTypes(formattedAggregates),
);

// TODO: decide if these should live in selectors or in eventual containers
export const mapStateToProps = createStructuredSelector({
  chartLoading: selectChartLoading,
  tableLoading: selectTableLoading,
  reasons: selectReasons,
  adminReasons: selectAdminReasons,
  aggregates: selectFormattedAggregates,
  categories: selectCategories,
  types: selectBandTypes,
  adminCategories: selectAdminCategories,
  reportOptions: selectReportOptions,
  bounceSearchOptions: selectReportSearchOptions,
});
