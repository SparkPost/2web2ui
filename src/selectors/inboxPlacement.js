import { createSelector } from 'reselect';
import endsWith from 'lodash/endsWith';

export const getSeeds = state => state.inboxPlacement.seeds;

export const selectReferenceSeed = createSelector([getSeeds], seeds =>
  seeds.find(seed => endsWith(seed, 'seed.sparkpost.com')),
);
