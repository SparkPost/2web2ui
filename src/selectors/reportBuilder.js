import { createSelector } from 'reselect';
import { formatAggregates } from 'src/helpers/bounce';
import _ from 'lodash';

const selectAggregates = ({ bounceReport }) => bounceReport.aggregates;

export const selectReasons = ({ bounceReport }) =>
  bounceReport.reasons.filter(reason => reason.count_bounce > 0);

export const selectFormattedAggregates = createSelector([selectAggregates], aggregates => {
  if (!aggregates.count_bounce && !aggregates.count_admin_bounce) {
    return [];
  }
  return formatAggregates(aggregates);
});
