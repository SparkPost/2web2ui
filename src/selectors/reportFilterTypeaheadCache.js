import { createSelector } from 'reselect';
import _ from 'lodash';

function reshape(list, type) {
  return list.map(value => ({ type, value }));
}

const selectSubaccounts = ({ subaccounts }) => subaccounts.list;
const selectSendingDomains = ({ sendingDomains }) => sendingDomains.list;
const selectRecipientDomains = ({ metrics }) => metrics.domains;
const selectCampaigns = ({ metrics }) => metrics.campaigns;
const selectSendingIps = ({ metrics }) => metrics.sendingIps;
const selectIpPools = ({ metrics }) => metrics.ipPools;
const selectTemplates = ({ metrics }) => metrics.templates;

const selectPreparedTemplates = createSelector([selectTemplates], templates =>
  reshape(templates, 'Template'),
);

const selectPreparedSubaccounts = createSelector([selectSubaccounts], subaccounts =>
  subaccounts.map(({ name, id }) => ({
    type: 'Subaccount',
    value: `${name} (ID ${id})`,
    id,
  })),
);

const selectPreparedSendingDomains = createSelector([selectSendingDomains], domains =>
  domains.map(d => ({ type: 'Sending Domain', value: d.domain })),
);

const selectPreparedRecipientDomains = createSelector([selectRecipientDomains], domains =>
  reshape(domains, 'Recipient Domain'),
);

const selectPreparedCampaigns = createSelector([selectCampaigns], campaigns =>
  reshape(campaigns, 'Campaign'),
);

const selectPreparedSendingIps = createSelector([selectSendingIps], ips =>
  reshape(ips, 'Sending IP'),
);

const selectPreparedIpPools = createSelector([selectIpPools], pools => reshape(pools, 'IP Pool'));

const selectCache = createSelector(
  [
    selectPreparedTemplates,
    selectPreparedSubaccounts,
    selectPreparedSendingDomains,
    selectPreparedRecipientDomains,
    selectPreparedCampaigns,
    selectPreparedSendingIps,
    selectPreparedIpPools,
  ],
  (...caches) => _.flatten(caches),
);

export const selectCacheReportBuilder = createSelector(
  [
    selectPreparedTemplates,
    selectPreparedSubaccounts,
    selectPreparedSendingDomains,
    selectPreparedRecipientDomains,
    selectPreparedCampaigns,
    selectPreparedSendingIps,
    selectPreparedIpPools,
  ],
  (templates, subaccounts, sendingDomains, recipientDomains, campaigns, sendingIps, IpPools) => ({
    'Recipient Domain': recipientDomains,
    Subaccount: subaccounts,
    Campaign: campaigns,
    Template: templates,
    'Sending Domain': sendingDomains,
    'IP Pool': IpPools,
    'Sending IP': sendingIps,
  }),
);

export default selectCache;
