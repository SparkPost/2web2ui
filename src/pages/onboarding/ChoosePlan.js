import React, { useEffect, useMemo } from 'react';
import { reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { Form } from 'src/components/tracking/form';
import { Grid, Button, Panel } from 'src/components/matchbox';
import { Heading } from 'src/components/text';
import { showAlert } from 'src/actions/globalAlert';
import { CenteredLogo, Loading, PlanPicker } from 'src/components';
import {
  getBillingCountries,
  verifyPromoCode,
  clearPromoCode,
  getPlans,
  getBundles,
} from 'src/actions/billing';
import billingCreate from 'src/actions/billingCreate';
import { choosePlanMSTP } from 'src/selectors/onboarding';
import promoCodeValidate from 'src/pages/billing/helpers/promoCodeValidate';
import { prepareCardInfo } from 'src/helpers/billing';
import PromoCodeNew from '../../components/billing/PromoCodeNew';
import { FORMS } from 'src/constants';
import CreditCardSection from './components/CreditCardSection';
import _ from 'lodash';
import { DASHBOARD_ROUTE } from 'src/constants';

export function OnboardingPlanPage({
  getPlans,
  getBundles,
  getBillingCountries,
  billingCreate,
  billingCreateLoading,
  billingCreateSuccess,
  showAlert,
  submitting,
  billing,
  verifyPromoCode,
  clearPromoCode,
  currentPlan,
  loading,
  selectedPlan = {},
  handleSubmit,
  hasError,
  bundles,
}) {
  const next_step = DASHBOARD_ROUTE;
  const history = useHistory();
  useEffect(() => {
    getPlans();
  }, [getPlans]);

  useEffect(() => {
    getBundles({ type: 'messaging' });
  }, [getBundles]);

  useEffect(() => {
    getBillingCountries();
  }, [getBillingCountries]);

  useEffect(() => {
    // if we can't get plans or countries form is useless
    // they can pick plan later from billing
    if (hasError) {
      history.push(next_step);
    }
  }, [next_step, hasError, history]);

  const isPlanFree = useMemo(() => Boolean(!_.get(selectedPlan, 'messaging.price', true)), [
    selectedPlan,
  ]);

  const onSubmit = values => {
    const selectedPromo = billing.selectedPromo;
    const newValues =
      values.card && !isPlanFree ? { ...values, card: prepareCardInfo(values.card) } : values;

    // no billing updates needed since they are still on free plan
    if (isPlanFree) {
      history.push(next_step);
      return;
    }
    const billingId = values.planpicker.messaging.billing_id;

    let action = Promise.resolve({});
    if (selectedPromo.promoCode && !isPlanFree) {
      const { promoCode } = selectedPromo;
      newValues.promoCode = promoCode;
      action = verifyPromoCode({
        promoCode,
        billingId,
        meta: { promoCode, showErrorAlert: false },
      });
    }

    // Note: billingCreate will update the subscription if the account is AWS
    return action.then(({ discount_id }) => {
      newValues.discountId = discount_id;
      return billingCreate({ ...newValues, billingId });
    });
  };

  const applyPromoCode = promoCode => {
    verifyPromoCode({
      promoCode,
      billingId: selectedPlan.messaging.billing_id,
      meta: { promoCode, showErrorAlert: false },
    });
  };

  const onPlanSelect = e => {
    if (currentPlan !== e.code) {
      clearPromoCode();
    }
  };

  const { selectedPromo = {}, promoError, promoPending } = billing;
  const promoCodeObj = {
    selectedPromo: selectedPromo,
    promoError: promoError,
    promoPending: promoPending,
  };
  const handlePromoCode = {
    applyPromoCode,
    clearPromoCode,
  };

  useEffect(() => {
    if (billingCreateSuccess) {
      history.push(next_step);
      showAlert({ type: 'success', message: 'Added your plan' });
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [billingCreateSuccess]);

  if (loading || !bundles.length || (!submitting && billingCreateLoading)) {
    return <Loading />;
  }

  const disableSubmit = submitting || promoPending;

  const buttonText = submitting ? 'Updating Subscription...' : 'Get Started';
  return (
    <Form onSubmit={handleSubmit(onSubmit)} id="onboarding-chooseplan-form">
      <CenteredLogo />
      <Grid>
        <Grid.Column>
          <Panel.LEGACY>
            <Panel.LEGACY.Section>
              <Heading as="h3">Select A Plan</Heading>
            </Panel.LEGACY.Section>
            <PlanPicker
              selectedPromo={selectedPromo}
              disabled={disableSubmit}
              bundles={bundles}
              onChange={onPlanSelect}
            />
            {!isPlanFree && (
              <Panel.LEGACY.Section>
                <PromoCodeNew
                  key={selectedPromo.promoCode || 'promocode'}
                  disabled={disableSubmit}
                  promoCodeObj={promoCodeObj}
                  handlePromoCode={handlePromoCode}
                />
              </Panel.LEGACY.Section>
            )}
            <CreditCardSection billing={billing} submitting={submitting} isPlanFree={isPlanFree} />
            <Panel.LEGACY.Section>
              <Button disabled={disableSubmit} type="submit" variant="primary">
                {buttonText}
              </Button>
            </Panel.LEGACY.Section>
          </Panel.LEGACY>
        </Grid.Column>
      </Grid>
    </Form>
  );
}

const formOptions = {
  form: FORMS.JOIN_PLAN,
  enableReinitialize: true,
  asyncValidate: promoCodeValidate(FORMS.JOIN_PLAN),
  asyncChangeFields: ['planpicker'],
  asyncBlurFields: ['promoCode'],
};

export default connect(choosePlanMSTP(FORMS.JOIN_PLAN), {
  billingCreate,
  showAlert,
  getPlans,
  getBillingCountries,
  verifyPromoCode,
  clearPromoCode,
  getBundles,
})(reduxForm(formOptions)(OnboardingPlanPage));
