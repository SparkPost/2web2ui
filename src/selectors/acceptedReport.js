import { createSelector } from 'reselect';
import { reshapeAttempts } from 'src/helpers/accepted';
import { computeKeysForItem } from 'src/helpers/metrics';
import { getAcceptedMetrics } from 'src/helpers/accepted';

const computeAccepted = computeKeysForItem(getAcceptedMetrics());

const selectAcceptedReport = (state) => state.acceptedReport;
const selectAttempts = createSelector(
  [selectAcceptedReport],
  ({ attempts }) => attempts
);
const selectAggregates = createSelector(
  [selectAcceptedReport],
  ({ aggregates }) => aggregates
);

export const selectAcceptedAttempts = createSelector(
  [selectAggregates, selectAttempts],
  (aggregates, attempts) => {
    if (!aggregates.count_accepted) {
      return [];
    }
    return reshapeAttempts(attempts);
  }
);

export const selectAcceptedAggregates = createSelector(
  [selectAggregates],
  (aggregates) => {
    if (!aggregates.count_accepted) {
      return [];
    }
    return computeAccepted(aggregates);
  }
);

