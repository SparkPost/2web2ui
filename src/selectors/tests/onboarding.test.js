import { choosePlanMSTP } from '../onboarding';
import { changePlanInitialValues } from 'src/selectors/accountBillingForms';

jest.mock('redux-form', () => ({
  formValueSelector: () => () => 'selected-plan-redux-form-value',
}));

jest.mock('src/selectors/accountBillingForms', () => ({
  changePlanInitialValues: jest.fn(() => 'change-plan-initial-values'),
}));

jest.mock('src/selectors/accountBillingInfo', () => ({
  selectAvailableBundles: jest.fn(() => []),
}));

describe('choosePlanMSTP', () => {
  let state;
  let props;

  beforeEach(() => {
    state = {
      account: {},
      billing: {},
      billingCreate: {
        loading: null,
        success: null,
        error: false,
      },
    };
    props = {
      location: {},
    };
  });

  it('should return mapped prop values', () => {
    const result = choosePlanMSTP()(state, props);
    expect(result).toMatchSnapshot();
    expect(result.loading).toEqual(false);
    expect(result.hasError).toEqual(false);
    expect(changePlanInitialValues).toHaveBeenCalledWith(state, { planCode: undefined });
  });

  it('should return when there is an error', () => {
    state.billing.bundlesError = true;
    const result = choosePlanMSTP()(state, props);
    expect(result.hasError).toEqual(true);
  });

  it('should return when loading is true', () => {
    state.account.loading = true;
    const result = choosePlanMSTP()(state, props);
    expect(result.loading).toEqual(true);
  });

  it('should pass on a plan code from location state', () => {
    props.location.state = {
      plan: 'passed-plan',
    };
    choosePlanMSTP()(state, props);
    expect(changePlanInitialValues).toHaveBeenCalledWith(state, { planCode: 'passed-plan' });
  });
});
