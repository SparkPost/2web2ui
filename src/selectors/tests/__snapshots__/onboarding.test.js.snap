// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choosePlanMSTP should return mapped prop values 1`] = `
Object {
  "billing": Object {},
  "billingCreateLoading": null,
  "billingCreateSuccess": null,
  "bundles": Array [],
  "hasError": false,
  "initialValues": "change-plan-initial-values",
  "loading": false,
  "selectedPlan": "selected-plan-redux-form-value",
}
`;
